Add dedupe middleware

dianashk opened the pull request:
- [x] Pad the result size to allow for duplicates to be removed while still returning the specified count.
- [x] Dedupe middleware removes __exact__ dupes and truncates the results
to the specified size.

Fixes pelias/api#339

dianashk commented:
@team, I'm not sure how to acceptance test this in our current framework. Any suggestions welcomed.

riordan commented:
I'd be fine with running manual acceptance tests on this one for now, provided we create a new, well-documented issue for how we'd update the framework.

If we do that, we should just document the testing methodology and results in detail here.
___________________________
We have dupes coming from a few places:
* Address duplication from OSM <> Open Addresses
* Landmark / POI duplication from OSM <> Geonames
* Admin area duplication from Geonames <> Quatroshapes

I welcome comments on this methodology. And the whole team should have edit access on this comment to rewrite it if we decide otherwise.


# Positive Testing:
## Addresses
The easiest to find + verify will be addresses. We pick 2-3 cities in OpenAddresses with great OSM address coverage (Chicago + NYC come to mind). We pick 5-10 arbitrary addresses in each, verify that we're putting out dupes now, and then see if it catches those exact matches in dev.

This is unlikely to to catch addresses where there's user-collected addresses in OSM (non-import) + OpenAddresses. I'd suspect there's going to be subtle variations in the addresses there. Maybe Amsterdam as a candidate here?

- [ ] 5-10 addresses w/ dupes now in Chicago
- [ ] 5-10 addresses w/ dupes now in NYC
- [ ] 5-10 addresses w/ dupes now in Amsterdam

## POI's
Landmarks between OSM <> Geonames _may_ be caught this way, but there's likely again to be variance in properties being compared.

- [ ] 5-10 childhood road trip destinations.

## Admin Areas
We might _actually_ catch a few of these since this doesn't take location into account and Quattroshapes imported a few of these from geonames.

- [ ] 10 regions, each in different countries

# Regressions
## Addresses
I'm actually having a hard time with this one

## POI's
Within the dataset there are a few identically named things that are actually different.
[Tower of London(focused from london)](http://pelias.github.io/compare/#/v1/search%3F&focus.point.lat=51.5081548283056&focus.point.lon=-0.12196540832519531&text=tower%20of%20london) is a great example (thx @missinglink for the updated slippy map in compare):
![pelias_geocoder_leaflet_plugin_and_abilene_-_google_maps](https://cloud.githubusercontent.com/assets/227472/11534721/2151ba7e-98de-11e5-94d2-32329a6405be.png)

We get the "Tower of London" from Geonames. And from OSM ([map](http://nominatim.openstreetmap.org/search.php?q=tower+of+london&viewbox=-112.5%2C49.16%2C112.5%2C-49.16)) we get, two bus stops named Tower of London, the actual Tower of London, and "Tower of London Hill".

Let's see how this does.

- [ ] Tower of London
- [ ] At least one other POI example

## Admin Areas
There are lots of similarly named admin areas worldwide. Some of them are the same/not the same. Lets see which Springfields we get (lot of the same) and San Fransisco's (city + county occupying the same area)
- [ ] Springfield (lots)
- [ ] San Fransisco (co-occupying space)

dianashk commented:
Still not covering cases like [this](http://pelias.github.io/compare/#/v1/search%3Ftext=30%20w%2026th%20street):

__query:__ `/v1/search?text=30 w 26th street`
__results:__

> 1)	30 West 26th Street, Manhattan, NY
 2)	30 West 26th Street, Merced, CA
 3)	30 West 26 Street, Manhattan, NY     <<<====== missing `th`, but otherwise same at (1)
 4)	3009 West 26th Street, Lawrence, KS
 5)	301 West 26th Street, Bradenton, FL
 6)	3005 West 26th Street, Lawrence, KS
 7)	3021 West 26th Street, Lawrence, KS
 8)	3001 West 26th Street, Lawrence, KS
 9)	3000 West 26th street, Lawrence, KS
10)	3000 West 26th Street, Lawrence, KS

dianashk commented:
Added simple normalizer for strings (remove punctuation and change to lowercase).

Running acceptance tests, we find 203 dupes in those results. Some queries have multiple sets of dupes in them, so the number is greater than the total number of tests in the suite.

stephenkhess commented on middleware/dedupe.js:
```diff
@@ -0,0 +1,95 @@
+  // loop through data items and only copy unique items to uniqueResults
+  var uniqueResults = [];
+
+  _.some(res.data, function (hit) {
```
that's pretty clever

missinglink replied:
you can use the native js versions instead of the `lodash` ones if you like, they were introduced in ECMA 5.1 and some old browsers haven't caught up *cough internut exploder* yet

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/some
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/every

dianashk replied:
I'd prefer to keep using the lodash module until it's supported everywhere.

stephenkhess commented:
:+1:

missinglink commented on middleware/dedupe.js:
```diff
@@ -0,0 +1,95 @@
+ */
+function propMatch(item1, item2, prop) {
+  if (normalizeString(item1[prop]) !== normalizeString(item2[prop])) {
+    throw 'different';
```
never seen this before ;)

dianashk replied:
is that good or bad? :confused:

missinglink replied:
it's not 'bad', just 'unusual'. the language allows it, so I guess it's valid :)

Wrapping it in an `Error` object would be the more 'usual' way of doing it, or even better creating a `ValidationError` class which extends from Error in order to differentiate between a sytnax error in the `propMatch` code vs. a logic error.

missinglink commented on middleware/dedupe.js:
```diff
@@ -0,0 +1,95 @@
+  if (!str) {
+    return '';
+  }
+  return _.words(str.toLowerCase()).join(' ');
```
this doesn't look to be safe for use with non ASCII encodings, worth testing.

you can probably replace it with something like `str.split(/[ ,]+/)`

normalizing strings like this is super tricky, its best to focus on removing the things you *don't* want rather than targeting the things you *do* want, otherwise it will end up reducing foreign languages to empty strings

missinglink replied:
I actually can't understand this `reComplexWord` regex https://github.com/lodash/lodash/blob/master/lodash.js#L204-L217 so maybe it works?

orangejulius replied:
This is fixed in #390

missinglink commented on query/reverse.js:
```diff
@@ -30,7 +31,7 @@ function generateQuery( clean ){
   // set size
   if( clean.size ){
-    vs.var( 'size', clean.size );
+    vs.var( 'size', calcSize(clean.size));
```
is it possible to call the `calcSize()` function as middleware or part of the sanitization steps?

it would be cool if the queries were not aware of any specific functionality and just needed to know about `clean`, it would mean that we could be free to enable/disable middleware without also having to change the queries.

dianashk replied:
good point. would a separate variable work? something like clean.querySize. it might be useful to have both for other middleware.

dianashk commented:
Logs from acceptance tests. query=undefined for reverse geocoding

> query = 	new york	||	hit = 	New York
 query = 	Paris	||	hit = 	Paris
 query = 	Paris	||	hit = 	Paris
 query = 	Simon-Dach-Straße 7	||	hit = 	1 Simon-Dach-Straße
 query = 	statue of liberty	||	hit = 	Statue de la Liberté
 query = 	statue of liberty	||	hit = 	Statue de la Liberté
 query = 	statue of liberty	||	hit = 	Statue de la Liberté
 query = 	statue of liberty	||	hit = 	Statue de la Liberté (réplique)
 query = 	statue of liberty	||	hit = 	Statue de Rouget de Lisle
 query = 	7 Simon-Dach-Straße	||	hit = 	1 Simon-Dach-Straße
 query = 	london	||	hit = 	London
 query = 	france	||	hit = 	France
 query = 	france	||	hit = 	France
 query = 	philadelphia	||	hit = 	Philadelphia
 query = 	union sq	||	hit = 	4 Union Square
 query = 	philadelphia, pa	||	hit = 	Philadelphia
 query = 	paris	||	hit = 	Paris
 query = 	paris	||	hit = 	Paris
 query = 	portland	||	hit = 	Portland
 query = 	30 West 26th Street	||	hit = 	30 West 26th Street
 query = 	30 West 26th Street	||	hit = 	30 West 26th Street
 query = 	30 West 26th Street	||	hit = 	3000 West 26th Street
 query = 	soho, new york	||	hit = 	Soho Square
 query = 	soho, new york	||	hit = 	Trump Soho
 query = 	statue of liberty	||	hit = 	Statue of Liberty Replica
 query = 	statue of liberty	||	hit = 	Statue of Liberty
 query = 	130 dean street brooklyn, ny	||	hit = 	130 Dean Street
 query = 	130 dean street brooklyn, ny	||	hit = 	130 Dean Street
 query = 	101 saint marks pl, new york	||	hit = 	101 Saint Marks Place
 query = 	101 saint marks pl, new york	||	hit = 	109 Saint Marks Place
 query = 	101 saint marks pl, new york	||	hit = 	10 Saint Marks Place
 query = 	101 saint marks pl, new york	||	hit = 	102 Saint Marks Place
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	30 West 26th Street
 query = 	undefined	||	hit = 	7 Washington Place
 query = 	undefined	||	hit = 	283 Mercer Street
 query = 	undefined	||	hit = 	8 Washington Place
 query = 	undefined	||	hit = 	15 Washington Place
 query = 	undefined	||	hit = 	5 Washington Place
 query = 	undefined	||	hit = 	3 Washington Place
 query = 	undefined	||	hit = 	7 Washington Place
 query = 	undefined	||	hit = 	283 Mercer Street
 query = 	undefined	||	hit = 	8 Washington Place
 query = 	undefined	||	hit = 	15 Washington Place
 query = 	undefined	||	hit = 	5 Washington Place
 query = 	undefined	||	hit = 	3 Washington Place
 query = 	undefined	||	hit = 	7 Washington Place
 query = 	undefined	||	hit = 	283 Mercer Street
 query = 	undefined	||	hit = 	8 Washington Place
 query = 	undefined	||	hit = 	15 Washington Place
 query = 	undefined	||	hit = 	5 Washington Place
 query = 	undefined	||	hit = 	3 Washington Place
 query = 	undefined	||	hit = 	7 Washington Place
 query = 	undefined	||	hit = 	283 Mercer Street
 query = 	undefined	||	hit = 	8 Washington Place
 query = 	undefined	||	hit = 	15 Washington Place
 query = 	undefined	||	hit = 	5 Washington Place
 query = 	undefined	||	hit = 	3 Washington Place
 query = 	ridgewood, ny	||	hit = 	Ridgewood
 query = 	ridgewood	||	hit = 	Ridgewood
 query = 	ridgewood	||	hit = 	Ridgewood
 query = 	williamsburg, ny	||	hit = 	Williamsburg
 query = 	williamsburg	||	hit = 	Williamsburg
 query = 	chelsea, ny	||	hit = 	Chelsea
 query = 	chelsea, ny	||	hit = 	Chelsea
 query = 	chelsea, ny	||	hit = 	Chelsea
 query = 	chelsea, ny	||	hit = 	Chelsea
 query = 	chelsea	||	hit = 	Chelsea
 query = 	chelsea	||	hit = 	Chelsea
 query = 	chelsea	||	hit = 	Chelsea
 query = 	chelsea	||	hit = 	Chelsea
 query = 	statue of liberty	||	hit = 	Statue of Liberty Replica
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	McDonald's, Central Singapore, Singapore	||	hit = 	McDonald's
 query = 	New Zealand	||	hit = 	New Zealand
 query = 	Arbil, Iraq	||	hit = 	Arbil
 query = 	Madrid, Spain	||	hit = 	Madrid
 query = 	Madrid, Spain	||	hit = 	Humanes de Madrid
 query = 	Madrid, Spain	||	hit = 	Madrid
 query = 	Madrid, Spain	||	hit = 	Madrid
 query = 	Madrid, Spain	||	hit = 	Madrid
 query = 	West Bengal, India	||	hit = 	West Bengal University of Technology
 query = 	West Bengal, India	||	hit = 	Tamluk, West Bengal Haldia Tamluk Mecheda Road
 query = 	30 West 26th Street, Manhattan, NY	||	hit = 	30 West 26th Street
 query = 	30 West 26th Street, Manhattan, NY	||	hit = 	30 West 26th Street
 query = 	30 West 26th Street, Manhattan, NY	||	hit = 	30 West 26th Street
 query = 	100 20th street	||	hit = 	100 20th Street
 query = 	100 20th street	||	hit = 	100 20th St
 query = 	Waiotapu	||	hit = 	Waiotapu
 query = 	Waiotapu	||	hit = 	Waiotapu Stream
 query = 	Beach Bablyon	||	hit = 	Bābli
 query = 	Beach Bablyon	||	hit = 	Bābli
 query = 	DiDi Dumpling	||	hit = 	Dumpling Mountain
 query = 	DiDi Dumpling	||	hit = 	Dumpling Creek
 query = 	union square	||	hit = 	Union Square
 query = 	union square	||	hit = 	Union Square
 query = 	union square	||	hit = 	Union Square
 query = 	union square	||	hit = 	Union Square
 query = 	union square	||	hit = 	2 Union Square
 query = 	union square	||	hit = 	4 Union Square
 query = 	union square	||	hit = 	Union Square
 query = 	union square	||	hit = 	Union Square
 query = 	DiDi dumpling	||	hit = 	Dumpling Mountain
 query = 	DiDi dumpling	||	hit = 	Dumpling Creek
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singarpur
 query = 	Singarpuram	||	hit = 	Singarpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	Singarpuram	||	hit = 	Singārpur
 query = 	wales	||	hit = 	Wales
 query = 	malmo	||	hit = 	Malmö
 query = 	new york	||	hit = 	New York
 query = 	london	||	hit = 	London
 query = 	madrid	||	hit = 	Madrid
 query = 	madrid	||	hit = 	Madrid
 query = 	madrid	||	hit = 	Madrid
 query = 	madrid	||	hit = 	Madrid
 query = 	madrid	||	hit = 	Madrid
 query = 	california	||	hit = 	California
 query = 	california	||	hit = 	California
 query = 	san francisco	||	hit = 	San Francisco
 query = 	california	||	hit = 	Callao
 query = 	Cardiff, GBR	||	hit = 	Cardiff
 query = 	Cardiff, GBR	||	hit = 	Cardiff Heights
 query = 	Cardiff, GBR	||	hit = 	Cardiff Heights
 query = 	Cardiff, GBR	||	hit = 	Cardiff
 query = 	Cardiff, GBR	||	hit = 	Cardiff South
 query = 	Hillside, Santa Clara	||	hit = 	Hillside
 query = 	Hillside, Santa Clara	||	hit = 	Hillside
 query = 	Hillside, Santa Clara	||	hit = 	Hillside
 query = 	Hillside, Santa Clara	||	hit = 	Hillside
 query = 	cardiff, gbr	||	hit = 	Cardiff
 query = 	cardiff, gbr	||	hit = 	Cardiff Heights
 query = 	cardiff, gbr	||	hit = 	Cardiff Heights
 query = 	cardiff, gbr	||	hit = 	Cardiff
 query = 	cardiff, gbr	||	hit = 	Cardiff South
 query = 	palo alto, ca	||	hit = 	Palo Alto
 query = 	palo alto	||	hit = 	Palo Alto
 query = 	crown heights	||	hit = 	Crown Heights
 query = 	crown heights	||	hit = 	Crown Heights
 query = 	mission, ca	||	hit = 	Mission Bay
 query = 	bronx	||	hit = 	Bronkhorstspruit
 query = 	Shepherd and Flock	||	hit = 	Shepherds Flock Academy
 query = 	Shepherd and Flock	||	hit = 	Shepherds Flock Church
 query = 	Shepherd and Flock	||	hit = 	French Southern And Antarctic Lands
 query = 	Shepherd and Flock	||	hit = 	French Southern And Antarctic Lands
 query = 	Shepherd and Flock	||	hit = 	French Southern And Antarctic Lands
 query = 	Shepherd and Flock	||	hit = 	French Southern And Antarctic Lands
 query = 	Shepherd and Flock	||	hit = 	Shepherd Creek
 query = 	22 Moor Park Lane	||	hit = 	22 Pastoor Moorkensstraat
 query = 	whole foods NY	||	hit = 	Whole Foods
 query = 	whole foods NY	||	hit = 	Whole Foods
 query = 	whole foods NY	||	hit = 	Whole Foods
 query = 	whole foods NY	||	hit = 	Whole Foods Market
 query = 	whole foods NY	||	hit = 	Whole Foods Market
 query = 	1 water st manhattan ny	||	hit = 	1 Water Street
 query = 	1 water st manhattan ny	||	hit = 	1 Water Street
 query = 	1 water st manhattan ny	||	hit = 	1 Water Street
 query = 	1 water st manhattan ny	||	hit = 	1 Water Street
 query = 	3122 16th St San Francisco, CA 94103	||	hit = 	3120 16th Street
 query = 	1 main st ny 11201	||	hit = 	1 Main Street
 query = 	1 main st ny 11201	||	hit = 	37 Main Street
 query = 	1 main st ny 11201	||	hit = 	40 Main Street
 query = 	1 main st ny 11201	||	hit = 	45 Main Street
 query = 	1 main st ny 11201	||	hit = 	16 Main Street
 query = 	101 saint mark pl 10009	||	hit = 	101 Saint Marks Place
 query = 	101 saint mark pl 10009	||	hit = 	101 Avenue A
 query = 	101 saint mark pl 10009	||	hit = 	101 Avenue B
 query = 	101 saint mark pl 10009	||	hit = 	109 Saint Marks Place
 query = 	30 w 26 st	||	hit = 	30 West 26th Street
 query = 	30 w 26 st	||	hit = 	30 West 26th Street
 query = 	30 w 26 st	||	hit = 	30 West 26th Street

dianashk commented on query/autocomplete.js:
```diff
@@ -30,9 +30,6 @@ function generateQuery( clean ){
 
-  // always 10 (not user definable due to caching)
-  vs.var( 'size', 10 );
-
```
removed this because it should be part of the defaults. now that defaults work properly this should just work.

dianashk commented on query/reverse.js:
```diff
@@ -22,15 +22,9 @@ function generateQuery( clean ){
-  vs.set({
-    'size': 1,
-    'boundary:circle:radius': '500km'
-  });
```
removed this because it should be part of the defaults. now that defaults work properly this should just work.

dianashk commented on query/reverse_defaults.js:
```diff
@@ -1,10 +1,10 @@
+var _ = require('lodash');
 
-module.exports = extend( false, peliasQuery.defaults, {
+module.exports = _.merge({}, peliasQuery.defaults, {
```
BUG FIX: `extend` will assign the resulting object to `peliasQuery.defaults`, thereby poisoning the clean well of default data for all subsequent uses of `peliasQuery.defaults`. See same fix in all 3 default files.

dianashk commented on routes/v1.js:
```diff
@@ -58,9 +60,11 @@ function addRoutes(app, peliasConfig) {
     search: createRouter([
       sanitisers.search.middleware,
       middleware.types,
+      middleware.calcSize(),
```
only needed for `/search` and `/reverse` as all other endpoints don't support `size` changes.

dianashk commented on test/unit/fixture/search_full_address.js:
```diff
@@ -1,6 +1,4 @@
-
-var peliasQuery = require('pelias-query'),
-    vs = new peliasQuery.Vars( peliasQuery.defaults );
+var vs = require('../../../query/search_defaults');
```
should be using search specific defaults, instead of common pelias defaults (see in other search fixtures as well)

dianashk commented:
- [x] moved calcSize out of query code and into a middleware
- [x] added `clean.querySize` variable to be used in query code instead of `clean.size`
- [x] fixed a bug related to query default loading because it was surfaced as part of this work

stephenkhess commented:
:+1:
